### PR TITLE
[UPD] Made the necessary changes to perform the experiments for pose_…

### PR DESCRIPTION
**Pull request by GerrySant**
The evaluation when making the appearance transfer is not very optimal, so there is room for improvement. At the moment, the performance decreases considerably depending on the number of appearances to be used for the meta prediction of each sample.

As soon as you have given me the go-ahead, I will launch the experiments.

### CODE REVIEW

File: args.py [hunk 1]
```diff
@@ -54,6 +71,24 @@
...
+                    help="Value to be used when spliting the amount of appearneces between train and validation.")
+
+# Add arguments for output paths
+parser.add_argument('--log_dir', type=str, default='logs', help='Path in which to save logs')
+parser.add_argument('--checkpoint_dir', type=str, default='checkpoints', help='Path in which to save checkpoints')
+
+# Add arguments for evaluation
+parser.add_argument('--pretrained_checkpoint', type=str, default='checkpoints', help='Path to the model to be evaluated.')
+parser.add_argument('--predictions_output', type=str, default='predictions.csv', help='Path to the file in which to save the results.')
+parser.add_argument('--test_appearances_ids', type=parse_int_list, default='76,118,784,418,995,478,425,273,967,610',
```
Comment (AmitMY) [inline args.py]: use `nargs="+", type=int` instead of `parse_int_list`

File: args.py [hunk 2]
```diff
@@ -12,6 +13,22 @@
 from pose_format.torch.representation.points import PointsRepresentation
 from pose_format.utils.reader import BufferReader
 
+def valid_appearance_option(value):
+    if value in ["from_splis_signers", "all_splits_appearances"]:
+        return value
+    elif os.path.isfile(value):
+        return value
```
Comment (AmitMY) [inline args.py:20]: why do you need a file?

File: args.py [hunk 3]
```diff
@@ -12,6 +13,22 @@
 from pose_format.torch.representation.points import PointsRepresentation
 from pose_format.utils.reader import BufferReader
 
+def valid_appearance_option(value):
+    if value in ["from_splis_signers", "all_splits_appearances"]:
```
Comment (AmitMY) [inline args.py]: typo `splis`

File: base_model.py [hunk 4]
```diff
@@ -1,26 +1,91 @@
+# from collections import Counter
```
Comment (AmitMY) [inline base_model.py]: delete the comented code, and then commit. then i will be able to see the changes. right now it is a lot of false "added"/"deleted"

File: data.py [hunk 5]
```diff
@@ -67,12 +69,15 @@ def __getitem__(self, index):
...
+    def __init__(self, data: List, is_train=False, anonymize=False, transfer_appearance=False, signers_poses={}):
         self.data = data
         self.is_train = is_train
+        self.anonymize = anonymize
+        self.transfer_appearance = transfer_appearance
+        self.signers_poses = signers_poses
 
     @staticmethod
-    def from_tfds(tf_dataset, pose_type: str, **kwargs):
+    def from_tfds(tf_dataset, pose_type: str, anonymize: bool = False, transfer_appearance: bool = False, **kwargs):
```
Comment (AmitMY) [inline data.py]: no need to add these arguments, they will be in `kwargs`

File: data.py [hunk 6]
```diff
@@ -108,7 +115,12 @@ def from_tfds(tf_dataset, pose_type: str, **kwargs):
                 "pose": pose,
                 "label": gloss_id if gloss_id > 0 else 0
             })
+            if datum["signer"].numpy() not in signers_poses:
+                signers_poses[datum["signer"].numpy()] = pose
 
+        kwargs['anonymize'] = anonymize
+        kwargs['transfer_appearance'] = transfer_appearance
+        kwargs['signers_poses'] = signers_poses
```
Comment (AmitMY) [inline data.py:119]: no need to overload

**Review by AmitMY — CHANGES_REQUESTED** [review]

File: args.py (hunk 2 above)
Comment (GerrySant) [reply args.py:20]: To load the poses with which to transfer the appearances (ex. 10k.pose).

File: args.py (hunk 1 above)
Comment (GerrySant) [reply args.py]: Resolved in the new commit.

File: args.py (hunk 3 above)
Comment (GerrySant) [reply args.py]: Resolved in the new commit.

File: base_model.py (hunk 4 above)
Comment (GerrySant) [reply base_model.py]: Resolved in the new commit.

File: data.py (hunk 5 above)
Comment (GerrySant) [reply data.py]: Resolved in the new commit.

File: data.py (hunk 6 above)
Comment (GerrySant) [reply data.py:119]: Resolved in the new commit.

File: args.py [hunk 7]
```diff
@@ -54,6 +63,25 @@
 parser.add_argument('--rep_distance', type=bool, default=True, help='use limb distance in the vectors?')
 parser.add_argument('--rep_angles', type=bool, default=True, help='use limb angle in the vectors?')
 
+# Pose Anonymization Arguments
+parser.add_argument('--anonymize', type=bool, default=False, help='If True, anonymizes the poses.')
+parser.add_argument('--transfer_appearance', type=bool, default=False, help='If True, randomly transfers the apperance of poses.')
+parser.add_argument('--appearances', type=valid_appearance_option, default="from_splits_signers", 
+                    help="Choose from 'from_splits_signers', 'all_splits_appearances', or provide a valid file path.")
+parser.add_argument('--validation_signers_fraction', type=float, default=0.1,
```
Comment (AmitMY) [inline args.py]: not clear to me why you need to split the appearances?
you take them all from the file, i assume, so it should be random

File: base_model.py [hunk 8]
```diff
@@ -5,22 +5,23 @@
...
-            "training": metrics.Accuracy(),
-            "validation": metrics.Accuracy(),
-            "test": metrics.Accuracy(),
-        }
+        # Metrics for training and validation
+        self.train_accuracy = Accuracy(num_classes=226, task="multiclass")
+        self.train_loss = MeanMetric()
+        self.val_accuracy = Accuracy(num_classes=226, task="multiclass")
+        self.val_loss = MeanMetric()
+        self.test_accuracy = Accuracy(num_classes=226, task="multiclass")
```
Comment (AmitMY) [inline base_model.py:25]: i would suggest you construct a dictionary, based on the split name:
```
self.metrics = {split: Accuracy(num_classes=226, task="multiclass") for split in ["train", "valid", "test"]}
```
and it seems like you use the loss for logging, so i would do the same
```
self.loss = {split:  MeanMetric() for split in ["train", "valid"]}
```

File: base_model.py [hunk 9]
```diff
@@ -29,95 +30,49 @@ def pred(self, *args):
...
-                    f.write(_id.decode('utf-8') + "," + str(p) + "\n")
-
-            with ZipFile('predictions.zip', 'w') as zipObj:
-                zipObj.write("predictions.csv")
-
-    # Training steps
+            loss += self.signer_loss * F.cross_entropy(signer_hat, signer)
+
+        # Using metrics for training and validation
+        if split == 'training':
```
Comment (AmitMY) [inline base_model.py]: generalize this using the dictionaries

File: base_model.py [hunk 10]
```diff
@@ -29,95 +30,49 @@ def pred(self, *args):
...
-
-        return {
-            "loss": loss,
-            "signer": signer,
-            "id": batch["id"],
-            "pred": torch.argmax(y_hat, dim=1),
-            "target": y
-        }
-
-    def step_end(self, split: str, outputs):
```
Comment (AmitMY) [inline base_model.py:53]: why are you no longer using `step_end` for logging, and `epoch_end` for a generic epoch handler?
less code, and easier to read in my opinion

File: data.py [hunk 11]
```diff
@@ -165,28 +165,32 @@ def get_autsl(split: str):
...
 def split_train_dataset(dataset: PoseClassificationDataset, ids):
     ids = set(ids)
     train_data = [d for d in dataset.data if d["signer"] not in ids]
     valid_data = [d for d in dataset.data if d["signer"] in ids]
+    anonymize = dataset.anonymize
+    transfer_appearance = dataset.transfer_appearance
+    signers_poses = dataset.signers_poses
 
-    return PoseClassificationDataset(train_data, is_train=True), PoseClassificationDataset(valid_data, is_train=False)
+    return PoseClassificationDataset(train_data, is_train=True, anonymize=anonymize, transfer_appearance=transfer_appearance, signers_poses=signers_poses), PoseClassificationDataset(valid_data, is_train=False, anonymize=anonymize, transfer_appearance=transfer_appearance, signers_poses=signers_poses)
```
Comment (AmitMY) [inline data.py:189]: i'd define `kwargs = namedtuple(anonymize=dataset.anonymize, transfer_appearance=dataset.transfer_appearance, signers_poses=dataset.signers_poses)` then just pass it for both as `**kwargs`

File: pred_debug.py [hunk 12]
```diff
@@ -0,0 +1,110 @@
+
```
Comment (AmitMY) [inline pred_debug.py]: what is this `debug` file? (same for train)
why not `pred` and `train`?

**Review by AmitMY — CHANGES_REQUESTED** [review]

File: args.py (hunk 7 above)
Comment (GerrySant) [reply args.py]: But it is to avoid that randomly, an appearance can appear in both train and validation.

File: base_model.py (hunk 10 above)
Comment (GerrySant) [reply base_model.py:53]: I saw that with pytorch lightning the way to do it was by using these three `on_train_epoch_end`, `on_validation_epoch_end`, `on_test_epoch_end`. https://lightning.ai/docs/pytorch/stable/common/lightning_module.html

I think I was running into some error with parent classes until I did this modification.

File: base_model.py (hunk 8 above)
Comment (GerrySant) [reply base_model.py:25]: Suggested modifications are applied in the next pushed commit.

File: base_model.py (hunk 9 above)
Comment (GerrySant) [reply base_model.py]: Suggested modifications are applied in the next pushed commit.

File: args.py (hunk 7 above)
Comment (AmitMY) [reply args.py]: but i think that is the point - that also in test, we have appearances that we see in training, no new appearances

File: data.py (hunk 11 above)
Comment (GerrySant) [reply data.py:189]: Suggested modifications are applied in the next pushed commit.

File: pred_debug.py (hunk 12 above)
Comment (GerrySant) [reply pred_debug.py]: Suggested modifications are applied in the next pushed commit.

File: base_model.py [hunk 13]
```diff
@@ -5,22 +5,20 @@
...
         self.signer_loss_patience = signer_loss_patience
 
-        self.metrics = {
-            "training": metrics.Accuracy(),
-            "validation": metrics.Accuracy(),
-            "test": metrics.Accuracy(),
-        }
+        # Metrics for training and validation
+        self.metrics = {split: Accuracy(num_classes=226, task="multiclass") for split in ["training", "validation", "test"]}
+        self.mean_loss = {split: MeanMetric() for split in ["training", "validation", "test"]}
```
Comment (AmitMY) [inline base_model.py]: wrap these two in `nn.ModuleDict(` then, you don't need to cast them with `.to`

File: base_model.py [hunk 14]
```diff
@@ -29,95 +27,47 @@ def pred(self, *args):
...
 
-    # Define steps
-
     def step(self, split: str, batch, batch_idx):
         y = batch["label"]
         signer = batch["signer"]
         y_hat, signer_hat = self(batch)
 
-        # self.logger.experiment.add_image('example_images', grid, 0)
+        self.metrics[split] = self.metrics[split].to(y_hat.device) if self.metrics[split].device != y_hat.device else self.metrics[split]
```
Comment (AmitMY) [inline base_model.py]: no need to cast to device, just use `nn.ModuleDict(`

File: args.py (hunk 7 above)
Comment (GerrySant) [reply args.py]: Do you prefer me to remove it? otherwise I think I prepared it to do not split the appearences if the parameter is set to 0

File: base_model.py [hunk 15]
```diff
@@ -29,95 +27,47 @@ def pred(self, *args):
...
+    def on_validation_epoch_end(self):
+        self.log('val_acc_epoch', self.metrics['validation'].compute())
+        self.log('val_loss_epoch', self.mean_loss['validation'].compute())
+        self.metrics['validation'].reset()
+        self.mean_loss['validation'].reset()
 
-    def validation_step_end(self, outputs):
-        return self.step_end("validation", outputs)
+    def on_test_epoch_end(self):
+        self.log('test_acc_epoch', self.test_accuracy.compute())
```
Comment (AmitMY) [inline base_model.py]: this parameter no longer exists. that's why i prefer you do a generic `on_epoch_end` like the original code

File: base_model.py [hunk 16]
```diff
@@ -29,95 +27,47 @@ def pred(self, *args):
...
 
-    # Validation steps
-
-    def validation_step(self, batch, batch_idx):
-        return self.step("validation", batch, batch_idx)
+    def on_validation_epoch_end(self):
+        self.log('val_acc_epoch', self.metrics['validation'].compute())
+        self.log('val_loss_epoch', self.mean_loss['validation'].compute())
+        self.metrics['validation'].reset()
+        self.mean_loss['validation'].reset()
```
Comment (AmitMY) [inline base_model.py]: duplication of code, should use a generic `on_epoch_end` like the original code

File: data.py (hunk 6 above)
Comment (AmitMY) [reply data.py:119]: seems like you are still overloading - you can just pass it as
```
 return PoseClassificationDataset(data,  signers_poses= signers_poses, **kwargs)
 ```

File: data.py (hunk 11 above)
Comment (AmitMY) [reply data.py:189]: i don't see that they are applied.

File: train.py [hunk 17]
```diff
@@ -20,39 +33,64 @@
...
 
-        callbacks.append(ModelCheckpoint(
-            filepath="models/" + logger.experiment.id + "/weights.ckpt",
-            verbose=True,
-            save_top_k=1,
-            monitor='validation_acc_epoch',
-            mode='max'
-        ))
+    num_gpus = torch.cuda.device_count()
+    devices = num_gpus if num_gpus > 0 else 1
```
Comment (AmitMY) [inline train.py]: `max(num_gpus, 1)`

File: train.py [hunk 18]
```diff
@@ -1,14 +1,27 @@
...
 
 # noinspection PyUnresolvedReferences
-from .args import args
-from .data import get_autsl, ZeroPadCollator
-from .model import PoseSequenceClassification
+from args import args
+from data import get_autsl, ZeroPadCollator
+from model import PoseSequenceClassification
+
+def extract_subdictionary(original_dict, fraction=0.1):
```
Comment (AmitMY) [inline train.py]: use the same appearances for train and validation, no need to split them

File: requirements.txt [hunk 19]
```diff
@@ -5,4 +5,4 @@ torch
 wandb
 tqdm
 pose_format
-git+git://github.com/sign-language-processing/datasets.git
+#git+git://github.com/sign-language-processing/datasets.git
```
Comment (AmitMY) [inline requirements.txt]: why?

**Review by AmitMY — CHANGES_REQUESTED** [review]

File: requirements.txt (hunk 19 above)
Comment (GerrySant) [reply requirements.txt]: I did it in order to avoid extra packages conflicts. I install it manually. However, i'll let it as it was in the next commit

File: base_model.py [hunk 20]
```diff
@@ -1,26 +1,27 @@
...
-        self.metrics = {
-            "training": metrics.Accuracy(),
-            "validation": metrics.Accuracy(),
-            "test": metrics.Accuracy(),
-        }
+        self.metrics = nn.ModuleDict()
+        for split in ["training", "validation", "test"]:
+            self.metrics[f"acc_{split}"] = Accuracy(num_classes=226, task="multiclass")
+            if split != "test":
+                self.metrics[f"mean_loss_{split}"] = MeanMetric()
```
Comment (AmitMY) [inline base_model.py:24]: your code before was a lot nicer.
just had to wrap it in a module dict
```py
splits = ["training", "validation", "test"]
self.metrics = nn.ModuleDict({split: Accuracy(num_classes=226, task="multiclass") for split in splits})
self.mean_loss = nn.ModuleDict({split: MeanMetric() for split in splits})
```

much cleaner and more declarative

File: base_model.py (hunk 10 above)
Comment (AmitMY) [reply base_model.py:53]: that is true, and to unify them all, the old code defined one method for each, then proxied it:
image

File: data.py (hunk 6 above)
Comment (AmitMY) [reply data.py:119]: resolved but you did not address it

File: data.py (hunk 11 above)
Comment (AmitMY) [reply data.py:189]: resolved but you did not address it

**Review by AmitMY — CHANGES_REQUESTED** [review]
